# GeekBrains Calculus. Lesson 2 & 3 Homework.

## Введение в математический анализ

1. **Как относятся друг к другу множество и последовательность? (в ответе использовать слова типа: часть, целое, общее, частное, родитель, дочерний субъект и т.д.)**

Последовательность - бесконечное упорядоченное множество элементов, которые могут повторяться.

2. **Прочитать высказывания математической логики, построить их отрицания и установить истинность.**

<center>
    $$
        1) \forall y \in [0;1]: sgn(y) = 1 \\
        2) \forall n \in \mathbb{N} > 2: \exists x,y,z \in \mathbb{N}: x^n = y^n + z^n \\
        3) \forall x \in \mathbb{R} \exists X \in \mathbb{R}: X > x \\
        4) \forall x \in \mathbb{C} \nexists y \in \mathbb{C} : x > y || x < y \\
        5) \forall y \in [0; \frac{\pi}{2}] \exists \epsilon > 0: sin y < sin(y + \epsilon) \\
        6) \forall y \in [0; \pi) \exists \epsilon > 0: cos y > cos(y + \epsilon) \\
        7) \exists x : x \notin \{\mathbb{N}, \mathbb{Z}, \mathbb{Q}, \mathbb{R}, \mathbb{C}\}
    $$
</center>

1) $ \forall y \in [0;1]: sgn(y) = 1 $ \
Чтение: Для всех чисел y в интервале от 0 до 1 включительно, знак положителен. \
Высказывание ложно. sgn(0) = 0. \
Отрицание - $ \exists y \in [0;1]: sgn(y) \neq 1 $.

2) $ \forall n \in \mathbb{N} > 2: \exists x,y,z \in \mathbb{N}: x^n = y^n + z^n $ \
Чтение: Для всех натуральных чисел n больше 2 существует хотя бы 1 набор натуральных чисел x, y, z, при котором x в степени n равно сумме y в степени n и z в степени n. \
Высказывание ложно. По последней теореме Ферма для n > 2 нет натуральных решений для такого уравнения. \
Отрицание - $ \exists n \in \mathbb{N} > 2: \forall x, y, z \in \mathbb{N}: x^n \neq y^n + z^n $.

3) $ \forall x \in \mathbb{R} \exists X \in \mathbb{R}: X > x $ \
Чтение: Для всех вещественных чисел x существует хотя бы 1 вещественное число X, которое больше x. \
Высказывание верно. Вещественные числа бесконечны, поэтому всегда есть число, большее, чем какое-либо. \
Отрицание - $ \exists x \in \mathbb{R} \forall X \in \mathbb{R}: X < x $.

4) $ \forall x \in \mathbb{C} \nexists y \in \mathbb{C} : x > y \| x < y $ \
Чтение: Для всех комплексных чисел x не существует такого комплексного числа y, при котором x больше или меньше y. \
Высказывание верно. Комплексные числа неупорядочены, поэтому мы не можем сказать, что одно строго больше или меньше другого - для этого нам потребуется определить, что значит быть больше/меньше. \
Отрицание - $ \exists x, y \in \mathbb{C} : x > y \| x < y $.

5) $ \forall y \in [0; \frac{\pi}{2}] \exists \varepsilon > 0: sin y < sin(y + \varepsilon) $ \
Чтение: Для всех чисел y в интервале от 0 до $\frac{\pi}{2}$ включительно существует такое положительное число $\varepsilon$, при котором $sin y$ меньше $sin(y + \varepsilon)$. \
Высказывание ложно. Если $y = \frac{\pi}{2} \Rightarrow sin y = 1$, что является максимальным значением функции синуса, а значит большего значения быть не может. \
Отрицание - $ \exists y \in [0; \frac{\pi}{2}] \forall \varepsilon > 0: sin y \ge sin(y + \varepsilon) $.

6) $ \forall y \in [0; \pi) \exists \varepsilon > 0: cos y > cos(y + \varepsilon) $ \
Чтение: Для всех чисел y в интервале от 0 включительно до $\pi$ невключительно существует такое положительное число $\varepsilon$, при котором $cos y$ больше $cos(y + \varepsilon)$. \
Высказывание верно. Добавив к y $\varepsilon: y + \varepsilon = \pi + 2\pi k$, можно получить $cos(y + \varepsilon) = cos \pi = -1$, что будет строго меньше любого $cos y, y \in [0; \pi)$. \
Отрицание - $ \exists y \in [0; \pi) \forall \varepsilon > 0: cos y \le cos(y + \varepsilon) $.

7) $ \exists x : x \notin \{\mathbb{N}, \mathbb{Z}, \mathbb{Q}, \mathbb{R}, \mathbb{C}\} $ \
Чтение: Существует хотя бы одно число x, не являющееся натуральным, целым, рациональным, действительным или комплексным. \
Высказывание верно. Есть числа, не вписывающиеся в комплексные - кардиналы, кватернионы. \
Отрицание - $ \forall x : x \in \{\mathbb{N}, \mathbb{Z}, \mathbb{Q}, \mathbb{R}, \mathbb{C}\} $.

## Множество

1. **Даны три множества a, b и с. Необходимо выполнить все изученные виды бинарных операций над всеми комбинациями множеств.**
2. **\*Выполнить задание 1 на языке Python.**

In [40]:
a = {1, 2, 3, 4, 5, 6, 7}
b = {1, 3, 7, 9, 10, 11, 13}
c = {2, 4, 8, 10, 12, 14}

Дополнения ко множеству (A) в Python нет - т.к. это все, что не во множестве A. \
Иначе: U-A, при этом U бесконечно.

**Объединение**

In [41]:
a | b

{1, 2, 3, 4, 5, 6, 7, 9, 10, 11, 13}

In [42]:
b | c

{1, 2, 3, 4, 7, 8, 9, 10, 11, 12, 13, 14}

In [43]:
a | c

{1, 2, 3, 4, 5, 6, 7, 8, 10, 12, 14}

In [44]:
a | b | c

{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14}

**Пересечение**

In [45]:
a & b

{1, 3, 7}

In [46]:
b & c

{10}

In [47]:
a & c

{2, 4}

In [48]:
a & b & c

set()

**Разность**

In [49]:
a - b

{2, 4, 5, 6}

In [50]:
b - c

{1, 3, 7, 9, 11, 13}

In [51]:
a - c

{1, 3, 5, 6, 7}

In [54]:
print(a - b - c)
print(a - c - b)
print(b - a - c)
print(b - c - a)
print(c - a - b)
print(c - b - a)

{5, 6}
{5, 6}
{9, 11, 13}
{9, 11, 13}
{8, 12, 14}
{8, 12, 14}


**Симметричная разность**

In [55]:
a ^ b

{2, 4, 5, 6, 9, 10, 11, 13}

In [56]:
b ^ c

{1, 2, 3, 4, 7, 8, 9, 11, 12, 13, 14}

In [57]:
a ^ c

{1, 3, 5, 6, 7, 8, 10, 12, 14}

In [58]:
print(a ^ b ^ c)
print(a ^ c ^ b)
print(b ^ a ^ c)
print(b ^ c ^ a)
print(c ^ a ^ b)
print(c ^ b ^ a)

{5, 6, 8, 9, 11, 12, 13, 14}
{5, 6, 8, 9, 11, 12, 13, 14}
{5, 6, 8, 9, 11, 12, 13, 14}
{5, 6, 8, 9, 11, 12, 13, 14}
{5, 6, 8, 9, 11, 12, 13, 14}
{5, 6, 8, 9, 11, 12, 13, 14}


## Последовательность

1. **Даны 4 последовательности. Необходимо:**
    - **исследовать их на монотонность;**
    - **исследовать на ограниченность;**
    - **найти пятый по счету член.**

<center>
    $$
        1) \left\{ a_n \right\}_{n=1}^{\infty} = 2^n - n \\
        2) \left\{ b_n \right\}_{n=2}^{\infty} = \frac{1}{1-n} \\
        3) \left\{ c_n \right\}_{n=1}^{\infty} = -1^n + \sqrt{2n} \\
        4) \left\{ d_n \right\}_{n=1}^{\infty} = (-1)^{2n} + \frac{1}{n^2}
    $$
</center>

1) $ \left\{ a_n \right\}_{n=1}^{\infty} = 2^n - n $
  - Последовательность возрастает.
  - Последовательность ограничена снизу: $ 1 \le \left\{ a_n \right\}_{n=1}^{\infty} < \infty $.
      - $2^n > n$ при любых n, поэтому наименьшее значение будет при наименьшем $n = 1 \Rightarrow a_1 = 2^1 - 1 = 1$.
      - Предел последовательности = $\infty$, т.к. $2^n$ растет быстрее $n$ и уходит в бесконечность, наибольшего значения нет.
  - 5-й член последовательности $a_5 = 27 $.

2) $ \left\{ b_n \right\}_{n=2}^{\infty} = \frac{1}{1-n} $
  - Последовательность возрастает.
  - Последовательность ограничена снизу и сверху: $ -1 \le \left\{ b_n \right\}_{n=2}^{\infty} < 0 $.
      - Минимальный член последовательности - $-1$, при $n = 2$.
      - Далее, с увеличением $n$, значение последовательности все растет, стремясь к $0$ при стремлении $n$ к $\infty$.
  - 5-й член последовательности $b_6 = -0.2$.

3) $ \left\{ c_n \right\}_{n=1}^{\infty} = -1^n + \sqrt{2n} $
  - Последовательность возрастает.
  - Последовательность ограничена снизу: $ -1 + \sqrt{2} \le \left\{ c_n \right\}_{n=1}^{\infty} < \infty $.
      - Функция $\sqrt{2n}$ растет до бесконечности, $-1^n$ не оказывает сильного влияния на нее.
      - $-1^n$ всегда равна -1.
      - $\sqrt{2n}$ минимальна при $n = 1$.
      - Значит, вся функция $-1^n + \sqrt{2n}$ минимальна при $n = 1$, ее минимальное значение = $-1 + \sqrt{2}$.
  - 5-й член последовательности $c_5 = -1 + \sqrt{10}$.

4) $ \left\{ d_n \right\}_{n=1}^{\infty} = (-1)^{2n} + \frac{1}{n^2} $
  - Последовательность убывает.
  - Последовательность ограничена снизу и сверху: $ 1 < \left\{ d_n \right\}_{n=1}^{\infty} \le 2 $.
      - $(-1)^{2n}$ всегда равно $1$.
      - $1/n^2$ максимально при $n = 1$ со значением $1$ и минимально при стремлении $n$ к $\infty$ со значением, стремящимся к 0.
  - 5-й член последовательности $d_5 = 1.04$.

2. **Найти 12-й член заданной неявно последовательности**

<center>
    $$ a_{1} = 128, a_{n+1} - a_{n} = 6 $$
</center>

In [59]:
n = 12
a = 128
for _ in range(n - 1):
    a += 6
a

194

Или же проще:

In [60]:
n = 12
a = 128 + 6 * (n - 1)
a

194

Ответ: $a_{12} = 194$.

3. **\*На языке Python предложить алгоритм вычисляющий численно предел с точностью $\varepsilon = 10^{-7}$.**

<center>
    $$ \lim_{n \to +\infty} \frac{n}{\sqrt[n]{n!}} $$
</center>

4. **\*Предложить оптимизацию алгоритма, полученного в задании 3, ускоряющую его сходимость.**